desktop_source/source-app.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+ * This file is part of wysiwidi
+ *
+ * Copyright (C) 2014 Intel Corporation.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA
+ */
+
+#include <iostream>
+#include <netinet/in.h> // htons()
+
+#include "source-app.h"
+
+void SourceApp::on_initialized(P2P::Client *client)
+{
+    scan();
+}
+
+void SourceApp::on_peer_added(P2P::Client *client, std::shared_ptr<P2P::Peer> peer)
+{
+    if (peer->device_type() == P2P::PRIMARY_SINK) {
+        peer->set_observer(this);
+    }
+}
+
+void SourceApp::on_peer_removed(P2P::Client *client, std::shared_ptr<P2P::Peer> peer)
+{
+    for (auto it = peers_.begin(); it != peers_.end(); ++it) {
```

Comment:
Really? I thought the iterator just gets invalidated (which is ok because we return immediately)... but it's possible that this code never got tested and is wrong.